### DISCUSSION TITLE
Experimenting with Django integration

### DISCUSSION THREAD

**Discussion by kkinder** [kkinder/puepy#42] [Show and tell]
For me, the plan all along has been that there could be a tidy backend integration. This is my first stab at it. The goal is that you have a frontend chunk and a backend chunk, closely coupled. Eg, from the demo:

```
class TodoView(PuePyView):
    frontend_url = "/static/todo.py"
    runtime = "mpy"

    @PuePyView.ajax
    def add_todo(self, item):
        todo_list.append(item)
        return todo_list
```

Then in the `static/todo.py` file:

```
self.state["todos"] = await backend.add_todo(self.state["todo_add"])
```

It uses json for serialization, with a little tweak to handle datetimes, so it's presumably pretty safe.

EDIT: I guess I forgot the link: https://github.com/kkinder/django-puepy

**Comment by kjbrawner22** [comment]
I took a different approach. Still working on it, but my goal was to have one module that contained the data models and their route mapping, and do it in a way that works with micropython. It's looking something like this (still working out details, would love feedback!):

my lib - `pyscript_api.py`
```python
class PyscriptAPI:
  # post and get decorators are defined here, other infrastructure
```

user models - `api_models.py`:
```python
from main import api

@api.request("/api/req", methods=["POST"])
class RequestModel:
  field_a: int
  field_b: str
  field_c: bool

# ...

@api.response("/api/req", methods=["GET", "POST"])
class ResponseModel:
  success: bool
  resp_field_1: str

```

**Reply by kkinder** [reply]
Yeah, that looks a lot like an API client. I wonder if you could actually make something that auto-generates classes like that based on [Django Ninja](https://django-ninja.dev) endpoints?

For my purposes, I was going for something closely resembling RPC, based on the theory that I want to be able to write a backend function and immediately be able to call it from the frontend without any additional hassles. Some time ago, I wrote a thing with a very similar spirit, but for pure JavaScript: https://github.com/kkinder/django-ajaxview

Anyway, I like what you're doing. Let me know if you get it working or have any puepy-side questions?
